### PR TITLE
check for running ghe-backup processes before spawn a new process

### PR DESCRIPTION
**Pull request by lotharschulz**
* approach focusses on pidof as suggested by ghe support / ronny

### CODE REVIEW

File: python/trigger_ghebackup.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import subprocess
+import sys
+
+
+def get_pid(name: str) -> list:
+    """Returns a list of pids, can be empty.
+    @attention: inspired by http://stackoverflow.com/a/26688998 on 09/02/2017
+    @author: Padraic Cunningham"""
+    if name:
+        return list(map(int, subprocess.check_output(["pidof", name]).split()))
```
Comment (jmcs) [inline python/trigger_ghebackup.py]: I would use a list comprehension here:

```python
return [int(pid) for pid in subprocess.check_output(["pidof", name]).split()]
```

since it makes it clearer from a cursory read that you are returning a list of integer PIDs

File: python/trigger_ghebackup.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
+import subprocess
+import sys
+
+
+def get_pid(name: str) -> list:
```
Comment (jmcs) [inline python/trigger_ghebackup.py]: You can use `-> List[int]` to specify that this will return a list only with integers. You can import `List` from `typing`

File: python/trigger_ghebackup.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    return []
+
+
+def start_ghe_backup(name: str) -> None:
+    """
+    Spawns process specified by given name
+    :param name: process to be spawned
+    """
+    try:
+        subprocess.Popen([name, '-v', '1>>', '/var/log/ghe-prod-backup.log', '2>&1'])
```
Comment (jmcs) [inline python/trigger_ghebackup.py]: You probably want to run this with `shell = True` and provide a string instead of a list, otherwise the output will not work.

You can also use [subprocess.run](https://docs.python.org/3/library/subprocess.html#subprocess.run):

```python
process = subprocess.run([name, '-v'], stdout=subprocess.PIPE,  stderr=subprocess.STDOUT)

with open('/var/log/ghe-prod-backup.log', 'a+') as log_file:
    log_file.write(process.output)
```

The second approach will block until the backup process is finished.

File: python/trigger_ghebackup.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+def trigger_backup(name: str, test: bool) -> str:
+    """
+    Check if there is a pid for the given name. If not, process after last space gets spawned.
+    :param name: process thats pid will be requested
+    :param test: does not start a process if false
+    :return: string either empty or the executable that shall be spawned
+    """
+    res = get_pid(name)
+    # process not running
+    if len(res) == 0:
```
Comment (jmcs) [inline python/trigger_ghebackup.py]: `if not res` is equivalent since empty lists are falsy.

File: python/trigger_ghebackup.py (hunk 3 above)
Comment (lotharschulz) [reply python/trigger_ghebackup.py]: Thanks for the input!

> You probably want to run this with shell = True and provide a string instead of a list, 
> otherwise the output will not work.

If the output works as intended remains unclear if tested with a test deployment.
That did not happen yet.

Based on 
* http://stackoverflow.com/a/3172488 
* http://stackoverflow.com/a/30831028

and similar resources on the web I decided _not_ to got for ```shell = True```.

File: .travis.yml [hunk 5]
```diff
@@ -9,9 +9,10 @@ before_install:
     - sudo apt-get install python3-dev
     - sudo apt-get install python3-yaml
     - sudo pip install boto3 --upgrade
-    - sudo pip install pyyaml --upgrade
+    #- sudo pip install pyyaml --upgrade
     - sudo pip install requests --upgrade
     - sudo pip install nose --upgrade
+    - sudo python -m pip install pyyaml
```
Comment (tuxlife) [inline .travis.yml]: Why not `pip install -r requirements.txt`?

File: .travis.yml [hunk 6]
```diff
@@ -9,9 +9,10 @@ before_install:
     - sudo apt-get install python3-dev
```
Comment (tuxlife) [inline .travis.yml]: Is it really necessary to install _dev_ packages with `apt-get`?

Comment (lotharschulz) [reply .travis.yml]: Was a [trial](https://github.com/zalando/ghe-backup/pull/43/commits/a2a125891c67229d747ce0162f405716934a888a) that was unfortunately not successful: https://travis-ci.org/zalando/ghe-backup/builds/201543299 the build is green, but the [ImportError](https://travis-ci.org/zalando/ghe-backup/builds/201543299#L382) should be fixed.
As of now, I don't have any idea why yaml package can't be imported.

Comment (tuxlife) [reply .travis.yml]: `extract_decrypt_kms.py` runs with `python3` and your pip maybe with `python2`

Comment (lotharschulz) [reply .travis.yml]: that should not be the case in travis context:
https://github.com/zalando/ghe-backup/blob/adaptBackupSchedule/.travis.yml#L4

File: .travis.yml (hunk 5 above)
Comment (lotharschulz) [reply .travis.yml]: thx, adapted ([3b6a9c7](https://github.com/zalando/ghe-backup/pull/43/commits/3b6a9c7f8f32c253d5fc08658877aa99ea2dbaa9), [d07dcd3](https://github.com/zalando/ghe-backup/pull/43/commits/d07dcd33453260da603a946b68fed5dae57848f5))

**Comment by lotharschulz** [comment]
issue ```binascii.Error: Incorrect padding``` (https://travis-ci.org/zalando/ghe-backup/builds/201891375)

**Comment by tuxlife** [comment]
Please use `bash` and not `sh`
`sh` != `bash`
https://travis-ci.org/zalando/ghe-backup/builds/201891375#L345

**Comment by tuxlife** [comment]
Are you sure, that you need a base64 encode before you decrypt things?

I think, the encrypted string are coded similar like baseXX

**Comment by lotharschulz** [comment]
> Are you sure, that you need a base64 encode before you decrypt things?
> I think, the encrypted string are coded similar like baseXX

@tuxlife : I am unsure , [you find examples either way](https://github.com/search?utf8=%E2%9C%93&q=kms+decrypt+language%3APython&type=Code&ref=searchresults).

Why would you change it?

File: .travis.yml [hunk 7]
```diff
@@ -1,21 +1,20 @@
 language: python
-
+sudo: required
```
Comment (tuxlife) [inline .travis.yml]: Why did you need **sudo**?

Comment (lotharschulz) [reply .travis.yml]: https://travis-ci.org/zalando/ghe-backup/builds/203975546/#L290 causes ```sudo required``` IMO. 
see also commit message of https://github.com/zalando/ghe-backup/pull/43/commits/f4c61e1d0910ce6a51caccf1296f118ea980a4df.

**Comment by lotharschulz** [comment]
could locally reproduce the padding issue:
```
...
Traceback (most recent call last):
  File "/Users/lschulz/zalando/dev/ghe-backup/python/test_extract_decrypt_kms.py", line 29, in test_several_aws_decrypt
    cls.test_aws_decrypt(cls.random_word(i))
  File "/Users/lschulz/zalando/dev/ghe-backup/python/test_extract_decrypt_kms.py", line 48, in test_aws_decrypt
    decryption_res = cls.kms.aws_decrypt(to_decrypt=encryption_res)
  File "/Users/lschulz/zalando/dev/ghe-backup/python/extract_decrypt_kms.py", line 56, in aws_decrypt
    CiphertextBlob=base64.b64decode(cls.decode_base64(data=to_decrypt))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/base64.py", line 90, in b64decode
    return binascii.a2b_base64(s)
nose.proxy.Error: Incorrect padding
...
```
that is similar to https://travis-ci.org/zalando/ghe-backup/builds/203982928/#L161 .

both:
* https://github.com/zalando/ghe-backup/pull/43/commits/016be0984d66e4072498ff9aef4abe8d50f06b08
* https://github.com/zalando/ghe-backup/pull/43/commits/a802d67184d8264f372c21bc432a24656b74f524 

fixed that locally.

**Comment by lotharschulz** [comment]
* test deployment performed
* instance up and running for several hours
* cron contains only
```
13 12,19 * * 7 root /trigger_ghebackup.py
```
* backup manually started via _trigger_ghebackup.py_ :
```
root@.....# /trigger_ghebackup.py
root@.....# Starting backup of xxxxx in snapshot 20170224T163423
Connect github.bus.zalan.do:122 OK (v2.8.7)
Backing up GitHub settings ...
* Transferring settings data ...
* Transferring license data ...
* Transferring management console password ...
* Transferring SAML keys ...
* Transferring UUID ...
Backing up SSH authorized keys ...
Backing up SSH host keys ...
Backing up MySQL database ...
...
```
next steps:
* wait for backup completed 
* test trigger_ghebackup.py once a backup process is running

note:
* padding issue e.g. https://github.com/zalando/ghe-backup/pull/43#issuecomment-281819728 did not happen

**Comment by lotharschulz** [comment]
trigger_ghebackup.py with active ghe-backup processes:
```
root        45  0.0  0.0  21100   616 ?        S    16:34   0:00 bash /backup/backup-utils/bin/ghe-backup -v 1>> /var/log/ghe-prod-b
root       658  0.0  0.1  21112  1856 ?        S    16:39   0:00  \_ bash /backup/backup-utils/share/github-backup-utils/ghe-backup-
root       787  0.0  0.2  21084  2828 ?        S    16:42   0:00      \_ bash /backup/backup-utils/share/github-backup-utils/ghe-rsy
root       794  0.0  0.1  21084  1440 ?        S    16:42   0:00          \_ bash /backup/backup-utils/share/github-backup-utils/ghe
root       797  0.1  1.8 175820 18844 ?        S    16:42   0:06          |   \_ rsync --bwlimit=4000 -av -e ghe-ssh -p 122 --link-d
root       800  2.5  0.5  49516  5544 ?        S    16:42   2:36          |   |   \_ ssh -p 22 -l admin -o StrictHostKeyChecking=no 
root       814  1.6  2.0 328280 21128 ?        S    16:42   1:38          |   |   \_ rsync --bwlimit=4000 -av -e ghe-ssh -p 122 --li
root       798  0.0  0.0  21084   364 ?        S    16:42   0:00          |   \_ bash /backup/backup-utils/share/github-backup-utils
root       799  0.0  0.0  14224   536 ?        S    16:42   0:00          |       \_ grep -E -v ^(file has vanished: |rsync warning:
root       795  0.0  0.1  21084  1700 ?        S    16:42   0:00          \_ bash /backup/backup-utils/share/github-backup-utils/ghe
root       796  0.0  0.0  14224   524 ?        S    16:42   0:00              \_ grep -E -v ^(file has vanished: |rsync warning: som
```
:

```
root@:/backup# /trigger_ghebackup.py
root@:/backup# bash /backup/backup-utils/bin/ghe-backup -v 1>> /var/log/ghe-prod-backup.log 2>&1
Error: backup process 45 of github.bus.zalan.do already in progress in snapshot 20170224T163423. Aborting.
exit
```

**Comment by lotharschulz** [comment]
@tuxlife : I added https://github.com/zalando/ghe-backup/pull/43/commits/29e950f27308e8a053439a9d177c3b8cef6a0334 similar to your start-backup-pidof suggestion.

File: start_backup.sh [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+# http://redsymbol.net/articles/unofficial-bash-strict-mode/
+set -euo pipefail
```
Comment (tuxlife) [inline start_backup.sh:3]: Very nice hint! :+1:

**Comment by tuxlife** [comment]
:+1:

**Comment by lotharschulz** [comment]
* test deployment was successful
* manual backup execution worked
* starting another backup approach did not start  another backup as expected
```
root@xxx:/backup# /start_backup.sh &
[1] 396
root@xxx:/backup# /start_backup.sh &
[2] 769
```

```
root@9ab8acc106a0:/backup# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
....
root       396  0.0  0.3  20980  3260 ?        S    17:19   0:00 /bin/bash /start_backup.sh
root       398  0.0  0.3  21096  3456 ?        S    17:19   0:00 bash /backup/backup-utils/bin/ghe-backup -v
....
root       772  0.0  0.3  37364  3236 ?        R+   17:22   0:00 ps aux
````

**Comment by lotharschulz** [comment]
waiting for backups triggered by cron tomorrow morning

**Comment by lotharschulz** [comment]
* yesterdays test deployment contain misconfigured cron - no backup triggered ;-(
* todays test deployment contains ```6 6,23 * * 1-6 root /start_backup.sh``` - should lead to a backup approach tomorrow.

**Comment by lotharschulz** [comment]
test deployment morning backup was successful:

screen shot 2017-03-02 at 08 58 09


```
lschulz@:~$ docker exec -it taupageapp bash
root@:/backup# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0  20980   300 ?        Ss   Mar01   0:00 /bin/bash /backup/final-docker-cmd.sh
root        14  0.0  0.0  29008   276 ?        Ss   Mar01   0:00 cron
root        15  0.0  0.0   7324    72 ?        S    Mar01   0:00 tail -F /var/log/ghe-prod-backup.log
root      1908  0.0  0.3  21160  3484 ?        Ss   07:50   0:00 bash
root      1922  0.0  0.3  37364  3196 ?        R+   07:50   0:00 ps aux
....
root@0ec6dbe466dc:/data/ghe-production-data# ll 20170301T230601
total 1660864
drwxr-xr-x 11 root root       4096 Mar  2 07:05 ./
drwxr-xr-x 10 root root       4096 Mar  2 07:09 ../
drwx------  3  500  500       4096 Oct 15 19:47 alambic_assets/
drwxr-xr-x  2 root root       4096 Mar  1 23:09 audit-log/
-rw-r--r--  1 root root       8359 Mar  1 23:06 authorized-keys.json
drwxr-xr-x  2 root root       4096 Mar  1 23:06 benchmarks/
drwx------  3  601  601       4096 Oct 16 18:47 elasticsearch/
-rw-r--r--  1 root root      17520 Mar  1 23:06 enterprise.ghl
drwxr-xr-x  3 root root       4096 Mar  2 06:46 git-hooks/
drwx------  2  500  500       4096 Oct 15 19:26 hookshot/
-rw-r--r--  1 root root          0 Mar  1 23:06 manage-password+
-rw-r--r--  1 root root 1026793943 Mar  1 23:08 mysql.sql.gz
drwx------ 10  500  500       4096 Oct 22  2015 pages/
-rw-r--r--  1 root root  673767435 Mar  1 23:08 redis.rdb
drwx------ 19  500  500       4096 Mar  1 23:12 repositories/
-rw-r--r--  1 root root      10240 Mar  1 23:06 saml-keys.tar
-rw-r--r--  1 root root      22699 Mar  1 23:06 settings.json
-rw-r--r--  1 root root      20480 Mar  1 23:06 ssh-host-keys.tar
drwx------ 18  500  500       4096 Mar 23  2016 storage/
-rw-r--r--  1 root root          6 Mar  1 23:06 strategy
-rw-r--r--  1 root root         37 Mar  1 23:06 uuid
-rw-r--r--  1 root root          7 Mar  1 23:06 version
```

**Comment by lotharschulz** [comment]
👍
